File: 6-api-edge-security-waf/2-csharp/WafMiddleware.cs
using System.Text;
using System.Text.Json;

namespace WafGateway;

// WAF middleware — the API-edge filter. It runs before the endpoint handlers,
// inspecting query string, headers, and JSON body. Malicious payloads are
// short-circuited with 403; clean requests flow on to the backend proxy.
public class WafMiddleware
{
    private readonly RequestDelegate _next;
    private readonly ILogger<WafMiddleware> _logger;

    public WafMiddleware(RequestDelegate next, ILogger<WafMiddleware> logger)
    {
        _next = next;
        _logger = logger;
    }

    public async Task InvokeAsync(HttpContext context)
    {
        // 1. Inspect every query-string value.
        foreach (var pair in context.Request.Query)
        {
            var hit = WafRules.InspectValue(pair.Value.ToString());
            if (hit != null) { await BlockAsync(context, hit); return; }
        }

        // 2. Inspect user-controlled headers.
        var headerHit =
            WafRules.InspectValue(context.Request.Headers["X-Forwarded-For"].ToString())
            ?? WafRules.InspectValue(context.Request.Headers["Referer"].ToString());
        if (headerHit != null) { await BlockAsync(context, headerHit); return; }

        // 3. Inspect the JSON body recursively, then rewind it for the handler.
        context.Request.EnableBuffering();
        if (context.Request.ContentLength is > 0)
        {
            using var reader = new StreamReader(
                context.Request.Body, Encoding.UTF8, leaveOpen: true);
            var raw = await reader.ReadToEndAsync();
            context.Request.Body.Position = 0;
            if (!string.IsNullOrEmpty(raw))
            {
                try
                {
                    using var doc = JsonDocument.Parse(raw);
                    var bodyHit = WafRules.InspectPayload(doc.RootElement);
                    if (bodyHit != null) { await BlockAsync(context, bodyHit); return; }
                }
                catch (JsonException) { /* non-JSON body: nothing to inspect */ }
            }
        }

        await _next(context);
    }

    private async Task BlockAsync(HttpContext context, string rule)
    {
        _logger.LogWarning("BLOCKED {Method} {Path} rule={Rule}",
            context.Request.Method, context.Request.Path, rule);
        // Uniform 403 contract — never echo the offending payload back.
        context.Response.StatusCode = StatusCodes.Status403Forbidden;
        await context.Response.WriteAsJsonAsync(new { status = "blocked", rule });
    }
}

File: 6-api-edge-security-waf/0-typescript/src/waf/waf.middleware.ts
import { Injectable, NestMiddleware, Logger } from "@nestjs/common"
import { Request, Response, NextFunction } from "express"
import { inspectPayload, inspectValue } from "./rules"

// WAF middleware — the API-edge filter. It runs BEFORE any route handler and
// inspects the query string, headers, and JSON body. A malicious payload is
// short-circuited with 403; a clean request is passed through to the backend.
@Injectable()
export class WafMiddleware implements NestMiddleware {
    private static readonly logger = new Logger("WAF")

    use(req: Request, res: Response, next: NextFunction): void {
        // 1. Inspect every query-string value.
        for (const value of Object.values(req.query)) {
            const hit = inspectValue(String(value))
            if (hit) return this.block(req, res, hit)
        }

        // 2. Inspect request headers that commonly carry user-controlled data.
        const headerHit =
            inspectValue(String(req.headers["x-forwarded-for"] ?? "")) ||
            inspectValue(String(req.headers["referer"] ?? ""))
        if (headerHit) return this.block(req, res, headerHit)

        // 3. Inspect the JSON body recursively (every leaf string value).
        const bodyHit = inspectPayload(req.body)
        if (bodyHit) return this.block(req, res, bodyHit)

        // Clean request — let it reach the route handler / backend proxy.
        next()
    }

    private block(req: Request, res: Response, rule: string): void {
        WafMiddleware.logger.warn(`BLOCKED ${req.method} ${req.path} rule=${rule}`)
        // Uniform 403 contract — never echo the offending payload back.
        res.status(403).json({ status: "blocked", rule })
    }
}
